repo2docker/contentproviders/zenodo.py

```diff
@@ -0,0 +1,99 @@
+import os
+import json
+import shutil
+
+from os import makedirs
+from os import path
+from urllib.request import urlopen, Request
+from zipfile import ZipFile, is_zipfile
+
+from .base import ContentProvider
+from ..utils import copytree
+
+
+class Zenodo(ContentProvider):
+    """Provide contents of a Zenodo deposit."""
+
+    def detect(self, doi, ref=None, extra_args=None):
+        doi = doi.lower()
+        # 10.5281 is the Zenodo DOI prefix
+        if doi.startswith("10.5281/"):
+            resp = urlopen("https://doi.org/{}".format(doi))
+            self.record_id = resp.url.rsplit("/", maxsplit=1)[1]
+            return {"record": self.record_id}
+
+        elif doi.startswith("https://doi.org/10.5281/") or doi.startswith(
+            "http://doi.org/10.5281/"
+        ):
+            resp = urlopen(doi)
+            self.record_id = resp.url.rsplit("/", maxsplit=1)[1]
+            return {"record": self.record_id}
+
+        elif doi.startswith("https://zenodo.org/record/") or doi.startswith(
+            "http://zenodo.org/record/"
+        ):
+            self.record_id = doi.rsplit("/", maxsplit=1)[1]
+            return {"record": self.record_id}
+
+    def fetch(self, spec, output_dir, yield_output=False):
+        record_id = spec["record"]
+
+        yield "Fetching Zenodo record {}.\n".format(record_id)
+        req = Request(
+            "https://zenodo.org/api/records/{}".format(record_id),
+            headers={"accept": "application/json"},
+        )
+        resp = urlopen(req)
+
+        record = json.loads(resp.read().decode("utf-8"))
+
+        def _fetch(file_ref, unzip=False):
```

Comment:
This function feels quite messy, but it needs to handle weird cases like "there is only one ZIP file in this zenodo record so lets unpack it, and if that generates a new sub directory move stuff from it to the top, then delete the sub dir and the zip file" and the variations on this.